src/Bicep.Decompiler/ArmHelpers/TemplateHelpers.cs

```diff
@@ -101,5 +126,142 @@ public static IEnumerable<JObject> FlattenAndNormalizeResource(JToken resourceJt
                 }
             }
         }
+
+        public static (JObject template, IReadOnlyDictionary<string, (FunctionExpression expression, string type)> parameters) ConvertNestedTemplateInnerToOuter(JObject template)
+        {
+            // this is useful to avoid naming clashes - we should prioritize names that have come from the parent template
+            var paramsAccessed = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            ExpressionHelpers.VisitExpressions(template, expression => 
+            {
+                if (expression is not FunctionExpression function || !IsFunction(function, "parameters"))
+                {
+                    return;
+                }
+
+                if (function.Parameters.Length != 1 || function.Parameters[0] is not JTokenExpression parameterName)
+                {
+                    throw new InvalidOperationException($"Found \"{function.Function}\" function with invalid signature");
+                }
+
+                paramsAccessed.Add(parameterName.Value.ToString());
+            });
+
+            // we don't want to populate this when we first visit parameters, because we may end up removing parameters
+            // - for example if they are only in-use inside 'reference()' functions
+            var parameters = new Dictionary<string, (FunctionExpression expression, string type)>(StringComparer.OrdinalIgnoreCase);
+
+            LanguageExpression ReplaceWithParameter(FunctionExpression function, string type, LanguageExpression paramNameExpression)
+            {
+                var withoutProperties = new FunctionExpression(
+                    function.Function,
+                    function.Parameters,
+                    Array.Empty<LanguageExpression>());
+                
+                var paramNameSerialized = ExpressionsEngine.SerializeExpression(paramNameExpression);
+                var paramName = UniqueNamingResolver.EscapeIdentifier(paramNameSerialized);
+
+                if (paramsAccessed.Contains(paramName))
+                {
+                    paramName = $"generated_{paramName}";
+                }
+
+                if (!parameters.ContainsKey(paramName))
+                {
+                    parameters[paramName] = (withoutProperties, type);
+                }
+
+                return new FunctionExpression(
+                    "parameters",
+                    new [] { new JTokenExpression(paramName) },
+                    function.Properties);
+            }
+
+            // process references first
+            template = ExpressionHelpers.RewriteExpressions(template, expression =>
+            {
+                if (expression is not FunctionExpression function || !IsFunction(function, "reference"))
```

Comment:
This is repeated a number of times, maybe it's worth extracting a helper method?